rust_src/src/buffers.rs

```diff
@@ -1057,6 +1060,51 @@ pub fn erase_buffer() {
     }
 }
 
+/// Return a string that is the name of no existing buffer based on NAME.
+/// If there is no live buffer named NAME, then return NAME.
+/// Otherwise modify name by appending `<NUMBER>', incrementing NUMBER
+/// (starting at 2) until an unused name is found, and then return that name.
+/// Optional second argument IGNORE specifies a name that is okay to use (if
+/// it is in the sequence to be tried) even if a buffer with that name exists.
+///
+/// If NAME begins with a space (i.e., a buffer that is not normally
+/// visible to users), then if buffer NAME already exists a random number
+/// is first appended to NAME, to speed up finding a non-existent buffer.
+#[lisp_fn(min = "1")]
+pub fn generate_new_buffer_name(name: LispStringRef, ignore: LispObject) -> LispStringRef {
+    if get_buffer(LispBufferOrName::Name(name.into())).is_none()
+        || (ignore != Qnil && string_equal(name.into(), ignore))
+    {
+        return name;
+    }
+
+    let basename = if name.byte_at(0) == b' ' {
+        let range = Range::new(0, 1000000);
+        let mut rng = rand::thread_rng();
+        let s = format!("-{}", range.ind_sample(&mut rng));
+        local_unibyte_string!(suffix, s);
+        let genname = unsafe { concat2(name.into(), suffix) };
+        if get_buffer(LispBufferOrName::Name(genname)).is_none() {
+            return genname.into();
+        }
+        genname
+    } else {
+        name.into()
+    };
+
+    let mut suffix_count = 2;
+    loop {
+        let s = format!("<{}>", suffix_count);
+        local_unibyte_string!(suffix, s);
+        let candidate = unsafe { concat2(basename, suffix) };
+        let buf = get_buffer(LispBufferOrName::Name(candidate));
+        if buf.is_none() || string_equal(candidate, ignore) {
```

Comment:
```suggestion
        if string_equal(candidate, ignore) || Fget_buffer(candidate).is_nil() {
```

Check for the string_equal first like in the C code. This will short circuit and avoid the call to `get_buffer`. Although not the creation of the `LispBufferOrName` object sadly. Which can be avoided if you use `Fget_buffer(candidate)` instead and check `is_nil()`.